File: app/security.py
"""Hashing de senha e emissão/validação de tokens JWT."""
import os
from datetime import datetime, timedelta, timezone
from typing import Optional

import bcrypt
from jose import jwt, JWTError

SECRET_KEY = os.getenv("JWT_SECRET", "dev-secret-change-me-in-production-please")
ALGORITHM = "HS256"
ACCESS_TOKEN_EXPIRE_MINUTES = 60 * 24  # 24h
REFRESH_TOKEN_EXPIRE_MINUTES = 60 * 24 * 7  # 7 dias


def hash_password(password: str) -> str:
    return bcrypt.hashpw(password.encode("utf-8")[:72], bcrypt.gensalt()).decode("utf-8")


def verify_password(plain: str, hashed: str) -> bool:
    try:
        return bcrypt.checkpw(plain.encode("utf-8")[:72], hashed.encode("utf-8"))
    except ValueError:
        return False


def _create_token(subject: str, role: str, expires_minutes: int, token_type: str) -> str:
    now = datetime.now(timezone.utc)
    payload = {
        "sub": subject,
        "role": role,
        "type": token_type,
        "iat": now,
        "exp": now + timedelta(minutes=expires_minutes),
    }
    return jwt.encode(payload, SECRET_KEY, algorithm=ALGORITHM)


def create_access_token(user_id: str, role: str) -> str:
    return _create_token(user_id, role, ACCESS_TOKEN_EXPIRE_MINUTES, "access")


def create_refresh_token(user_id: str, role: str) -> str:
    return _create_token(user_id, role, REFRESH_TOKEN_EXPIRE_MINUTES, "refresh")


def decode_token(token: str) -> Optional[dict]:
    try:
        return jwt.decode(token, SECRET_KEY, algorithms=[ALGORITHM])
    except JWTError:
        return None

File: src/utils/security.js
'use strict';
/**
 * Hashing de senha e emissão/validação de tokens JWT.
 * Equivalente a app/security.py.
 */
const bcrypt = require('bcryptjs');
const jwt = require('jsonwebtoken');

const SECRET_KEY = process.env.JWT_SECRET || 'dev-secret-change-me-in-production-please';
const ALGORITHM = 'HS256';
const ACCESS_TOKEN_EXPIRE_MINUTES = 60 * 24; // 24h
const REFRESH_TOKEN_EXPIRE_MINUTES = 60 * 24 * 7; // 7 dias

function hashPassword(password) {
  // bcrypt trunca em 72 bytes, igual ao comportamento do backend original.
  const truncated = Buffer.from(password, 'utf-8').slice(0, 72).toString('utf-8');
  return bcrypt.hashSync(truncated, 10);
}

function verifyPassword(plain, hashed) {
  try {
    const truncated = Buffer.from(plain, 'utf-8').slice(0, 72).toString('utf-8');
    return bcrypt.compareSync(truncated, hashed);
  } catch (err) {
    return false;
  }
}

function _createToken(subject, role, expiresMinutes, tokenType) {
  const payload = { sub: subject, role, type: tokenType };
  return jwt.sign(payload, SECRET_KEY, {
    algorithm: ALGORITHM,
    expiresIn: `${expiresMinutes}m`,
  });
}

function createAccessToken(userId, role) {
  return _createToken(userId, role, ACCESS_TOKEN_EXPIRE_MINUTES, 'access');
}

function createRefreshToken(userId, role) {
  return _createToken(userId, role, REFRESH_TOKEN_EXPIRE_MINUTES, 'refresh');
}

function decodeToken(token) {
  try {
    return jwt.verify(token, SECRET_KEY, { algorithms: [ALGORITHM] });
  } catch (err) {
    return null;
  }
}

module.exports = {
  hashPassword,
  verifyPassword,
  createAccessToken,
  createRefreshToken,
  decodeToken,
};
